### ISSUE TITLE
wine: support on Darwin

### ISSUE THREAD

**Issue by Pamelloes** [NixOS/nixpkgs#8849]
The wine expression builds an ELF binary--incompatible with OS X. I'm not too familiar with the technical details related so I don't really have much more to say. :/ I can provide more details if necessary.

**Comment by vcunat** [comment]
Currently wine on darwin seems unsupported in nixpkgs.

**Comment by Profpatsch** [comment]
(triage) still the case?

**Comment by justinweiss** [comment]
Yep, I think so -- trying to install 1.9.14, I get a:

```
error: a ‘i686-linux’ is required to build ‘/nix/store/psadbd27c471w4sjqwf3mldz1w81iyhw-wine-1.9.14.drv’, but I am a ‘x86_64-darwin’
```

**Comment by Profpatsch** [comment]
Yes, it’s restricted to the linux platforms. Can you try removing all platform restrictions from the wine nix files (i.e. `packages.nix` and the assert in `base.nix` and try to build and run? Maybe it works without any hassle.

**Comment by justinweiss** [comment]
I'm still kind of muddling through Nix, so I could be missing something, but it looks like there are still some problems. 

I had the same error after removing the assert and adding x86_64-darwin to the platforms -- as far as I can tell, it looks like the `pkgsi686Linux` forces the platform to be `i686-linux`. Changing that line to `pkgs` fails, because `cups` is a dependency (I'm guessing I'd have similar problems with alsa, etc.)

It looks like Wine on Darwin doesn't require too many dependencies anymore -- I wonder if disabling a bunch of the optional dependencies would get it working.

**Comment by Profpatsch** [comment]
I’m not sure how to do proper cross-compiling on my linux system.

    nix-build -E 'with import <nixpkgs> { system = "x86_64-darwin"; }; wine'

seems to build just fine, but it looks like that uses an `x86_64-linux` stdenv and then compiles to darwin somehow.@vcunat can I simulate a true darwin env from my machine?

**Comment by justinweiss** [comment]
Disabling most of the optional dependencies, I got to the point where it'll start to build wine, but fails: 

```
~/Source/nixpkgs[master *] jweiss$ nix-env -i wine-1.9.14 -f default.nix
installing ‘wine-1.9.14’
these derivations will be built:
  /nix/store/p9g170ky9fck4nh5pzd76hnvkcnp0dji-wine-1.9.14.drv
building path(s) ‘/nix/store/8ky2vx01x1m9jyzqizd13c04pydk30pz-wine-1.9.14’
unpacking sources
unpacking source archive /nix/store/vcr5177phiz4vb0vjsg4kg62bz18pvqr-wine-1.9.14.tar.bz2
source root is wine-1.9.14
setting SOURCE_DATE_EPOCH to timestamp 1467989999 of file wine-1.9.14/tools/wrc/writeres.c
patching sources
configuring
configure flags: --prefix=/nix/store/8ky2vx01x1m9jyzqizd13c04pydk30pz-wine-1.9.14  
checking build system type... x86_64-apple-darwin15.6.0
checking host system type... x86_64-apple-darwin15.6.0
checking whether make sets $(MAKE)... yes
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether clang accepts -g... yes
checking for clang option to accept ISO C89... none needed
checking whether we are using the GNU C++ compiler... yes
checking whether clang++ accepts -g... yes
checking for cpp... cpp
checking for ld... ld
checking whether clang -m32 works... no
configure: error: Cannot build a 32-bit program, you need to install 32-bit development libraries.
```

It looks like on the Linux side, this was solved with `callPackage_i686` and `pkgsi686Linux`, to force the system to build 32-bit versions of the dependencies. As far as I can tell, a non-64-bit Darwin platform isn't defined in `platforms.nix`.

**Comment by Profpatsch** [comment]
Is wine 32bit-only?

**Comment by justinweiss** [comment]
I think most people will build both 32-bit and 64-bit wine, because 64-bit builds will only run 64-bit software. There's a WoW64 thing that allows both kinds of software to be run, but still requires both 32- and 64-bit builds: https://wiki.winehq.org/Building_Wine#Shared_WoW64

**Comment by vcunat** [comment]
@Profpatsch: default wine is forced to `i686-linux`; your passing won't override that. Cross-compilation isn't that easy with nixpkgs.

**Comment by Profpatsch** [comment]
@vcunat Can wine generally not be built as x64?

**Comment by vcunat** [comment]
It can be. It can run 64-bit and mixed `*.exe`s, but our default is 32-bit only, as it covers most use cases and the mixed one has a really huge closure.

**Comment by bendlas** [comment]
wine's mixed 32/64 (wow) build first builds the 64bit version and then compiles 32bit binaries with a flag that makes them call out to their 64bit pendants on encountering win64 code.
So a full set of wine's dependencies is required as 32bit code, hence the reliance on  pkgsI686Linux on x86_64.

wine's 64bit support is a long shot from its 32bit mode and not considered well supported. Also, suprisingly few windows applications seem to be win64, even today. Probably because everybody is transitioning to .net anyway.

Bottom line, even if wine64/wineWow works on OSX, wine32 is the one we'd want to have working first.

Can anybody speak to compiling 32bit code on OSX? If we don't have an equivalent to pkgsI686Linux, maybe we can make one to support wine's dependencies?

**Comment by bendlas** [comment]
Apparently, wine64 on mac isn't happening any time soon: http://wine.1045685.n5.nabble.com/Wine-64-bit-tp5788608p5788773.html

wine32 on darwin would be valuable, though.

**Comment by vcunat** [comment]
I think we might best create "i686-darwin" (pseudo-)platform to support such use case(s).

**Comment by veprbl** [comment]
I was able to build wineMinimal with:
```diff
diff --git a/pkgs/misc/emulators/wine/base.nix b/pkgs/misc/emulators/wine/base.nix
index 66ea152..9a53489 100644
--- a/pkgs/misc/emulators/wine/base.nix
+++ b/pkgs/misc/emulators/wine/base.nix
@@ -3,14 +3,12 @@
   pkgconfig, fontforge, makeWrapper, flex, bison,
   supportFlags,
   buildScript ? null, configureFlags ? ""
+  ,llvmPackages_39,darwin
 }:
 
-assert stdenv.isLinux;
-assert stdenv.cc.cc.isGNU or false;
-
 with import ./util.nix { inherit lib; };
 
-stdenv.mkDerivation ((lib.optionalAttrs (! isNull buildScript) {
+llvmPackages_39.stdenv.mkDerivation ((lib.optionalAttrs (! isNull buildScript) {
   builder = buildScript;
 }) // rec {
   inherit name src configureFlags;
@@ -21,6 +19,7 @@ stdenv.mkDerivation ((lib.optionalAttrs (! isNull buildScript) {
 
   buildInputs = toBuildInputs pkgArches (with supportFlags; (pkgs:
   [ pkgs.freetype ]
+  ++ (with darwin.apple_sdk.frameworks; [CoreServices Cocoa ForceFeedback]) ++ [ darwin.CF ]
   ++ lib.optional pngSupport             pkgs.libpng
   ++ lib.optional jpegSupport            pkgs.libjpeg
   ++ lib.optional cupsSupport            pkgs.cups
diff --git a/pkgs/misc/emulators/wine/packages.nix b/pkgs/misc/emulators/wine/packages.nix
index a786bab..2578e45 100644
--- a/pkgs/misc/emulators/wine/packages.nix
+++ b/pkgs/misc/emulators/wine/packages.nix
@@ -5,22 +5,15 @@
 
 let src = lib.getAttr wineRelease (callPackage ./sources.nix {});
 in with src; {
-  wine32 = callPackage_i686 ./base.nix {
+  wine32 = callPackage ./base.nix {
     name = "wine-${version}";
     inherit src version supportFlags;
-    pkgArches = [ pkgsi686Linux ];
-    geckos = [ gecko32 ];
-    monos =  [ mono ];
-    platforms = [ "i686-linux" "x86_64-linux" ];
-  };
-  wine64 = callPackage ./base.nix {
-    name = "wine64-${version}";
-    inherit src version supportFlags;
     pkgArches = [ pkgs ];
     geckos = [ gecko64 ];
     monos =  [ mono ];
     configureFlags = [ "--enable-win64" ];
-    platforms = [ "x86_64-linux" ];
+    #configureFlags = [ "--without-pthread" "--without-x" "--without-freetype" ];
+    platforms = [ "x86_64-linux" "x86_64-darwin" ];
   };
   wineWow = callPackage ./base.nix {
     name = "wine-wow-${version}";
```
However the resulting wine segfaults on start. When I remove ```--enable-win64``` I able to get through configure until it fails during make:
```
ld: warning: ignoring file /nix/store/hv9riig5pilb7fh8gmk2sn70nlj7srr1-Libsystem-osx-10.11.6/lib/libSystem.dylib, file was built for x86_64 which is not the architecture being linked (i386): /nix/store/hv9riig5pilb7fh8gmk2sn70nlj7srr1-Libsystem-osx-10.11.6/lib/libSystem.dylib
ld: warning: ignoring file /nix/store/k3dvpyc0h334drbbh8kk5cwxy4fyrysf-clang-3.9.1/bin/../lib/clang/3.9.1/lib/darwin/libclang_rt.osx.a, missing required architecture i386 in file /nix/store/k3dvpyc0h334drbbh8kk5cwxy4fyrysf-clang-3.9.1/bin/../lib/clang/3.9.1/lib/darwin/libclang_rt.osx.a (2 slices)
Undefined symbols for architecture i386:
<snip>
```
I looked into libSystem and it's problem is that it gets "re-exported" in [pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix](https://github.com/NixOS/nixpkgs/blob/b69f568f4c3ebaf48a7f66b0f051d28157a61afb/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix)
```
% file /nix/store/hv9riig5pilb7fh8gmk2sn70nlj7srr1-Libsystem-osx-10.11.6/lib/libSystem.dylib
/nix/store/hv9riig5pilb7fh8gmk2sn70nlj7srr1-Libsystem-osx-10.11.6/lib/libSystem.dylib: Mach-O 64-bit dynamically linked shared library x86_64                                                                               
% file /usr/lib/libSystem.dylib 
/usr/lib/libSystem.dylib: Mach-O universal binary with 2 architectures
/usr/lib/libSystem.dylib (for architecture i386):	Mach-O dynamically linked shared library i386
/usr/lib/libSystem.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
```
Perhaps this can be resolved with bringing universal binaries back into the picture somehow.

**Comment by veprbl** [comment]
libSystem issue is down after doing:
```diff
diff --git a/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix b/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
index 92c54d9..b0a42f3 100644
--- a/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
+++ b/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
@@ -76,6 +76,12 @@ appleDerivation rec {
        /usr/lib/libSystem.dylib \
        -reexported_symbols_list ${./system_symbols}
 
+    cat ${./system_c_symbols} | egrep -v "(INODE64|NOCANCEL|_off32|_off64|_skip)" > system_c_symbols_i386
+    ld -macosx_version_min 10.7 -arch i386 -dylib \
+       -o $out/lib/system/libsystem_i386.dylib \
+       /usr/lib/libSystem.dylib \
+       -reexported_symbols_list system_c_symbols_i386
+
     libs=$(otool -arch x86_64 -L /usr/lib/libSystem.dylib | tail -n +3 | awk '{ print $1 }')
 
     for i in $libs; do
@@ -85,7 +91,7 @@ appleDerivation rec {
     done
 
     ld -macosx_version_min 10.7 -arch x86_64 -dylib \
-       -o $out/lib/libSystem.B.dylib \
+       -o $out/lib/libSystem.B.x86_64.dylib \
        -compatibility_version 1.0 \
        -current_version 1226.10.1 \
        -reexport_library $out/lib/system/libsystem_c.dylib \
@@ -93,6 +99,8 @@ appleDerivation rec {
        -reexport_library $out/lib/libSystem_internal.dylib \
        $args
 
+    lipo -create $out/lib/libSystem.B.x86_64.dylib $out/lib/system/libsystem_i386.dylib -output $out/lib/libSystem.B.dylib
+
     ln -s libSystem.B.dylib $out/lib/libSystem.dylib
 
     # Set up links to pretend we work like a conventional unix (Apple's design, not mine!)
```
I now get lesser amount of unresolved symbols:
```
ld: warning: ignoring file /nix/store/iaj3c86hv7ja98zhpc65agfmli9v2q9b-clang-3.9.1/bin/../lib/clang/3.9.1/lib/darwin/libclang_rt.osx.a, missing required architecture i386 in file /nix/store/iaj3c86hv7ja98zhpc65agfmli9v2q9b-clang-3.9.1/bin/../lib/clang/3.9.1/lib/darwin/libclang_rt.osx.a (2 slices)
Undefined symbols for architecture i386:
  "___bzero", referenced from:
      _parse_makefile in makedep-6dfae3.o
  "___error", referenced from:
      _create_dir in makedep-6dfae3.o
      _parse_file in makedep-6dfae3.o
      _create_temp_file in makedep-6dfae3.o
      _rename_temp_file in makedep-6dfae3.o
  "_fdopen$UNIX2003", referenced from:
      _create_temp_file in makedep-6dfae3.o
  "_fopen$UNIX2003", referenced from:
      _main in makedep-6dfae3.o
      _open_input_makefile in makedep-6dfae3.o
      _load_file in makedep-6dfae3.o
  "_fputs$UNIX2003", referenced from:
      _main in makedep-6dfae3.o
  "_free", referenced from:
      _get_expanded_make_variable in makedep-6dfae3.o
      _strmake in makedep-6dfae3.o
      _create_dir in makedep-6dfae3.o
      _parse_file in makedep-6dfae3.o
      _open_local_file in makedep-6dfae3.o
      _open_global_header in makedep-6dfae3.o
      _output_sources in makedep-6dfae3.o
      ...
  "_fwrite$UNIX2003", referenced from:
      _main in makedep-6dfae3.o
      _fatal_error in makedep-6dfae3.o
      _fatal_perror in makedep-6dfae3.o
  "_getpid", referenced from:
      _create_temp_file in makedep-6dfae3.o
  "_malloc", referenced from:
      _main in makedep-6dfae3.o
      _get_relative_path in makedep-6dfae3.o
      _parse_makefile in makedep-6dfae3.o
      _get_expanded_make_variable in makedep-6dfae3.o
      _get_line in makedep-6dfae3.o
      _strmake in makedep-6dfae3.o
      _add_generated_source in makedep-6dfae3.o
      ...
  "_mkdir", referenced from:
      _create_dir in makedep-6dfae3.o
  "_open$UNIX2003", referenced from:
      _create_temp_file in makedep-6dfae3.o
  "_realloc", referenced from:
      _main in makedep-6dfae3.o
      _set_make_variable in makedep-6dfae3.o
      _get_line in makedep-6dfae3.o
      _strmake in makedep-6dfae3.o
      _add_dependency in makedep-6dfae3.o
      _add_include in makedep-6dfae3.o
      _parse_pragma_directive in makedep-6dfae3.o
      ...
  "_rename", referenced from:
      _rename_temp_file in makedep-6dfae3.o
  "_unlink", referenced from:
      _main in makedep-6dfae3.o
      _cleanup_files in makedep-6dfae3.o
      _rename_temp_file in makedep-6dfae3.o
```
These perhaps should come from clang.

**Comment by veprbl** [comment]
Next iteration:
```diff
diff --git a/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix b/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
index 92c54d9..ff54fcf 100644
--- a/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
+++ b/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
@@ -76,6 +76,12 @@ appleDerivation rec {
        /usr/lib/libSystem.dylib \
        -reexported_symbols_list ${./system_symbols}
 
+    cat ${./system_c_symbols} | egrep -v "(INODE64|NOCANCEL|_off32|_off64|_skip)" > system_c_symbols_i386
+    ld -macosx_version_min 10.7 -arch i386 -dylib \
+       -o $out/lib/system/libsystem_c.i386.dylib \
+       /usr/lib/libSystem.dylib \
+       -reexported_symbols_list system_c_symbols_i386
+
     libs=$(otool -arch x86_64 -L /usr/lib/libSystem.dylib | tail -n +3 | awk '{ print $1 }')
 
     for i in $libs; do
@@ -85,7 +91,7 @@ appleDerivation rec {
     done
 
     ld -macosx_version_min 10.7 -arch x86_64 -dylib \
-       -o $out/lib/libSystem.B.dylib \
+       -o $out/lib/libSystem.B.x86_64.dylib \
        -compatibility_version 1.0 \
        -current_version 1226.10.1 \
        -reexport_library $out/lib/system/libsystem_c.dylib \
@@ -93,6 +99,15 @@ appleDerivation rec {
        -reexport_library $out/lib/libSystem_internal.dylib \
        $args
 
+    ld -macosx_version_min 10.7 -arch i386 -dylib \
+       -o $out/lib/libSystem.B.i386.dylib \
+       -compatibility_version 1.0 \
+       -current_version 1226.10.1 \
+       -reexport_library $out/lib/system/libsystem_c.i386.dylib \
+       $args
+
+    lipo -create $out/lib/libSystem.B.x86_64.dylib $out/lib/libSystem.B.i386.dylib -output $out/lib/libSystem.B.dylib
+
     ln -s libSystem.B.dylib $out/lib/libSystem.dylib
 
     # Set up links to pretend we work like a conventional unix (Apple's design, not mine!)
```
```
Undefined symbols for architecture i386:
  "___error", referenced from:
      _parse_file in makedep-fbe712.o
      _create_dir in makedep-fbe712.o
      _create_temp_file in makedep-fbe712.o
      _rename_temp_file in makedep-fbe712.o
  "_fdopen$UNIX2003", referenced from:
      _create_temp_file in makedep-fbe712.o
  "_fopen$UNIX2003", referenced from:
      _open_input_makefile in makedep-fbe712.o
      _load_file in makedep-fbe712.o
      _rename_temp_file_if_changed in makedep-fbe712.o
  "_fwrite$UNIX2003", referenced from:
      _output_dependencies in makedep-fbe712.o
      _fatal_error in makedep-fbe712.o
      _fatal_perror in makedep-fbe712.o
  "_getpid", referenced from:
      _create_temp_file in makedep-fbe712.o
  "_mkdir", referenced from:
      _create_dir in makedep-fbe712.o
  "_open$UNIX2003", referenced from:
      _create_temp_file in makedep-fbe712.o
  "_rename", referenced from:
      _rename_temp_file in makedep-fbe712.o
  "_unlink", referenced from:
      _cleanup_files in makedep-fbe712.o
      _rename_temp_file in makedep-fbe712.o
      _rename_temp_file_if_changed in makedep-fbe712.o
ld: symbol(s) not found for architecture i386
```

**Comment by knedlsepp** [comment]
I'm playing around with this a little. Everything compiles, but exits with a segfault:
https://github.com/knedlsepp/nixpkgs/blob/fix-wine-on-darwin/pkgs/misc/emulators/wine/darwin.nix

**Comment by veprbl** [comment]
@knedlsepp Did you check if wine64 works on darwin at all?

**Comment by knedlsepp** [comment]
Well, not in detail. But it looks like you can get it via homebrew: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wine.rb#L445

**Comment by bendlas** [comment]
Sorry if you already read that, but I'd be surprised if wine64 on mac could run actual 64bit code, even if it builds: http://wine.1045685.n8.nabble.com/Wine-64-bit-td5788608.html#a5788642

Regarding the segfault: The [guide](https://wiki.winehq.org/MacOS/Building) mentions building with `clang++`. I don't know if that's implicit in Nix' darwin stdenv, but I don't see it in your build recipe. Did you get the 32bit build to work @knedlsepp?

**Comment by knedlsepp** [comment]
@bendlas: According to [this thread](https://github.com/Homebrew/legacy-homebrew/issues/47428) on legacy-homebrew the ABI problem might be resolved with the newer version of clang. So I think we shouldn't loose all hope. With regards to the stdenv: clang is the default on darwin.

**Comment by bendlas** [comment]
The thread I linked mentioned registers being overwritten _by OSX_, for legacy reasons. That's not a thing to be fixed with a compiler update, because wine doesn't recompile (or otherwise emulate) windows binaries, it just links them up correctly and runs them natively.

The thread you linked doesn't mention running win64 apps sucessfully either. Can you try out the current version in homebrew with an actual win64 binary? e.g http://www.chip.de/downloads/Notepad2-64-Bit_48895401.html

**Comment by knedlsepp** [comment]
Just tried via:
```
brew install wine
wine64 ~/Downloads/Notepad2_4.2.25_x64.exe
wine64 ~/.wine/drive_c/Program\ Files/Notepad2/Notepad2.exe
```
And it seems to run just fine:
![asdf](https://user-images.githubusercontent.com/3287933/29590758-b47e20e6-879b-11e7-8e1e-ae1682a5f3f8.png)

**Comment by bendlas** [comment]
O.O

I stand corrected. Ironically, this doesn't even currently work on linux: https://github.com/NixOS/nixpkgs/issues/28486
I'll work on that in the meantime.

Holler when you get something to test. I'll fire up my hackintosh vm for the occasion ;-)

**Comment by knedlsepp** [comment]
After stepping through the segfault with gdb it turns out there is some kind of locale problem.
`LC_ALL=C wine64 ~/Downloads/notepad++.exe` produces the following situation. The error message saying it misses the FreeType fonts can also be seen in the screenshot. Seems putting it in the buildInputs is not quite enough.

![notepad++](https://user-images.githubusercontent.com/3287933/29740612-bd5563ca-8a5a-11e7-8fce-39c89b9376e0.png)

**Comment by veprbl** [comment]
@matthewbauer Is it too early to close this? I think win32 applications are not supported yet.
